source/extensions/filters/http/grpc_stats/config.cc

```diff
@@ -0,0 +1,135 @@
+namespace Extensions {
+namespace HttpFilters {
+namespace GrpcStats {
+
+namespace {
+
+class GrpcStatsFilter : public Http::StreamFilter {
+public:
+  explicit GrpcStatsFilter(Grpc::Context& context) : context_(context) {}
+
+  // Http::StreamFilterBase
+  void onDestroy() override {}
+
+  // Http::StreamDecoderFilter
+  Http::FilterHeadersStatus decodeHeaders(Http::HeaderMap& headers, bool) override {
+    grpc_request_ = Grpc::Common::hasGrpcContentType(headers);
+    if (grpc_request_) {
+      cluster_ = decoder_callbacks_->clusterInfo();
+      if (cluster_) {
+        request_names_ = context_.resolveServiceAndMethod(headers.Path());
+      }
+    }
+    return Http::FilterHeadersStatus::Continue;
+  }
+  Http::FilterDataStatus decodeData(Buffer::Instance& data, bool) override {
+    if (grpc_request_) {
+      uint64_t delta = IncrementMessageCounter(data, &request_counter_);
+      if (delta > 0) {
+        maybeWriteState();
+        if (doStatTracking()) {
+          context_.chargeRequestStat(*cluster_, *request_names_, delta);
+        }
+      }
+    }
+    return Http::FilterDataStatus::Continue;
+  }
+  Http::FilterTrailersStatus decodeTrailers(Http::HeaderMap&) override {
+    return Http::FilterTrailersStatus::Continue;
+  }
+  void setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callbacks) override {
+    decoder_callbacks_ = &callbacks;
+  }
+
+  // Http::StreamEncoderFilter
+  Http::FilterHeadersStatus encode100ContinueHeaders(Http::HeaderMap&) override {
+    return Http::FilterHeadersStatus::Continue;
+  }
+  Http::FilterHeadersStatus encodeHeaders(Http::HeaderMap& headers, bool end_stream) override {
+    grpc_response_ = Grpc::Common::isGrpcResponseHeader(headers, end_stream);
+    if (doStatTracking()) {
+      context_.chargeStat(*cluster_, Grpc::Context::Protocol::Grpc, *request_names_,
+                          headers.GrpcStatus());
+    }
+    return Http::FilterHeadersStatus::Continue;
+  }
+  Http::FilterDataStatus encodeData(Buffer::Instance& data, bool) override {
+    if (grpc_response_) {
+      uint64_t delta = IncrementMessageCounter(data, &response_counter_);
+      if (delta > 0) {
+        maybeWriteState();
+        if (cluster_ && request_names_) {
+          context_.chargeResponseStat(*cluster_, *request_names_, delta);
+        }
+      }
+    }
+    return Http::FilterDataStatus::Continue;
+  }
+  Http::FilterTrailersStatus encodeTrailers(Http::HeaderMap& trailers) override {
+    if (doStatTracking()) {
+      context_.chargeStat(*cluster_, Grpc::Context::Protocol::Grpc, *request_names_,
+                          trailers.GrpcStatus());
+    }
+    return Http::FilterTrailersStatus::Continue;
+  }
+  Http::FilterMetadataStatus encodeMetadata(Http::MetadataMap&) override {
+    return Http::FilterMetadataStatus::Continue;
+  }
+  void setEncoderFilterCallbacks(Http::StreamEncoderFilterCallbacks&) override {}
+
+  void maybeWriteState() {
+    if (filter_object_ == nullptr) {
```

Comment:
Can you add a config option which defaults to false for tracking per-filter state? There is no reason for the normal user to pay for this perf-wise. Only Istio cares about this.

Reply:
We are not using this in istio yet. I'm OK refusing altogether the filter state for now, until istio finds a way to plug-into the stats pipeline here. It might work better via gRPC stat context, for example. However, I'm not sure why you think this has significant perf implications. I store the raw pointer in the filter to the filter data to avoid constant look-ups into the filter state.

Reply:
Lets just kill this for now if Istio is not going to use this. No matter what you do it has non-zero cost to create the object for each stream, etc. At very high RPS in the hot path we should be trying to optimize away needless allocations, etc.

Reply:
Killed it.

Reply:
Brought it back with a filter config defaulting to disabling it.